gpytorch/lazy/diag_lazy_tensor.py

```diff
@@ -3,159 +3,131 @@
 
+from itertools import product
 import torch
+from .added_diag_lazy_tensor import AddedDiagLazyTensor
 from .lazy_tensor import LazyTensor
+from ..utils.broadcasting import _mul_broadcast_shape
 
 
 class DiagLazyTensor(LazyTensor):
-    def __init__(self, diag, exact_root_decomposition=False):
+    def __init__(self, diag):
         """
-        Diagonal lazy tensor
+        Diagonal lazy tensor. Supports arbitrary batch sizes.
 
         Args:
-            - diag (tensor) (batch) diagonal of matrix
-            - exact_root_decomposition (bool, default False) return an exact root decomposition.
-                Set to True only if the represented diagonal is relatively small.
+            :attr:`diag` (Tensor):
+                A `b1 x ... x bk x n` Tensor, representing a `b1 x ... x bk`-sized batch
+                of `n x n` diagonal matrices
         """
         super(DiagLazyTensor, self).__init__(diag)
         self._diag = diag
-        self.exact_root_decomposition = exact_root_decomposition
+        self._batch_shape = diag.shape[:-1]
 
-    def _matmul(self, rhs):
-        if rhs.ndimension() == 1 and self.ndimension() == 2:
-            return self._diag * rhs
-        else:
-            res = self._diag.unsqueeze(-1).expand_as(rhs) * rhs
-            return res
+    def __add__(self, other):
+        if isinstance(other, DiagLazyTensor):
+            return self.add_diag(other._diag)
+        return AddedDiagLazyTensor(other, self)
 
-    def _t_matmul(self, rhs):
-        # Matrix is symmetric
-        return self._matmul(rhs)
+    def _get_indices(self, indices):
+        """Extract elements from the LazyTensor. Supports arbitrary batch sizes and broadcasting across batch sizes.
+
+        Args:
+            :attr:`indices` (LongTensor):
+                A `b1 x ... bk x 2 x j` LongTensor, for each (multi-dimensional) batch selecting j
+                elements from the associated `n x n` diagonal matrix. The `2 x j` tensor `t` associated
+                with the last two dimensions represents the indices as lists of row and column indices,
+                that is, selecting the `b1 x ... x bk x t(0, l) x t(1, l)`-th element.
+                The batch indices follow standard broadcasting rules.
+
+        """
+        batch_shape = _mul_broadcast_shape(indices.shape[:-2], self._diag.shape[:-1])
+        row_indices = indices[..., 0, :].expand(batch_shape + indices.shape[-1:])
+        diag = self._diag.expand(batch_shape + self._diag.shape[-1:])
+
+        diag_elements = torch.stack(
+            [diag[batch_idx][row_indices[batch_idx]] for batch_idx in product(*(range(i) for i in batch_shape))]
+        ).view(batch_shape + indices.shape[-1:])
+
+        equal_indices = (indices[..., -1, :] == indices[..., -2, :]).type_as(self._diag)
+        return diag_elements * equal_indices
+
+    def _matmul(self, rhs):
+        # this are trivial if we multiply two DiagLazyTensors
+        if isinstance(rhs, DiagLazyTensor):
```

Comment:
I'd prefer if this check appeared in an overwritten `matmul` instead of in `_matmul` directly (I feel that `_matmul` should always return tensors). On the subject, I think that `LazyTensor.matmul` should by default return a `MatmulLazyTensor` if it gets called with a `LazyTensor` right hand side instead of erroring.

Reply:
moved to a new `.matmul` method, which just calls the superclass's method if `rhs` is not a `DiagLazyTensor`